File: Src/Test/TextAnalysis.Test/CommandLexerTests.cs
namespace Citron;

using Citron.LexicalAnalysis;
using System;
using System.Collections.Generic;
using System.IO;
using System.Text;
using System.Threading.Tasks;
using Xunit;

using static Citron.TextAnalysisTestMisc;

public class CommandLexerTests
{
    IEnumerable<Token> Process(Lexer lexer, LexerContext context)
    {
        var result = new List<Token>();

        while(!context.Pos.IsReachEnd())
        {
            var lexResult = lexer.LexCommandMode(context);
            if (!lexResult.HasValue) break;

            context = lexResult.Context;
            result.Add(lexResult.Token);
        }

        return result;
    }

    LexerContext RepeatLexNormal(List<Token> tokens, Lexer lexer, LexerContext context, bool bSkipNewLine, int repeatCount)
    {
        for (int i = 0; i < repeatCount; i++)
        {
            var result = lexer.LexNormalMode(context, bSkipNewLine);
            tokens.Add(result.Token); // ps
            context = result.Context;
        }

        return context;
    }

    LexerContext RepeatLexCommand(List<Token> tokens, Lexer lexer, LexerContext context, int repeatCount)
    {
        for (int i = 0; i < repeatCount; i++)
        {
            var result = lexer.LexCommandMode(context);
            tokens.Add(result.Token); // ps
            context = result.Context;
        }

        return context;
    }
            

    [Fact]
    public void TestLexerProcessStringExpInCommandMode()
    {
        var lexer = new Lexer();
        var context = MakeLexerContext("  p$$s${ ccc } \"ddd $e  \r\n }");

        var tokens = new List<Token>();

        context = RepeatLexCommand(tokens, lexer, context, 2);
        context = RepeatLexNormal(tokens, lexer, context, false, 2);
        context = RepeatLexCommand(tokens, lexer, context, 6);
            
        var expectedTokens = new Token[]
        {
            new TextToken("  p$s"),
            Tokens.DollarLBrace,
            new IdentifierToken("ccc"),
            Tokens.RBrace,
            new TextToken(" \"ddd "),
            new IdentifierToken("e"),
            new TextToken("  "),
            Tokens.NewLine,
            new TextToken(" "),
            Tokens.RBrace,
        };

        Assert.Equal(expectedTokens, tokens, TokenEqualityComparer.Instance);
    }

    [Fact]
    public void TestCommandModeLexCommands()
    {
        var lexer = new Lexer();
        var context = MakeLexerContext("ls -al");

        var result = Process(lexer, context);

        var expectedTokens = new Token[]
        {
            new TextToken("ls -al")
        };

        Assert.Equal(expectedTokens, result, TokenEqualityComparer.Instance);
    }

    [Fact]
    public void TestLexMultilines()
    {
        var lexer = new Lexer();
        var context = MakeLexerContext(@"
hello world \n

    hello    

}");
        var tokens = new List<Token>();
        RepeatLexCommand(tokens, lexer, context, 6);

        var expected = new List<Token>
        {
            Tokens.NewLine,
            new TextToken("hello world \\n"), Tokens.NewLine, // skip multi newlines
            new TextToken("    hello    "), Tokens.NewLine,                
            Tokens.RBrace
        };

        Assert.Equal(expected, tokens, TokenEqualityComparer.Instance);
    }

    [Fact]
    public void TestCommandModeLexCommandsWithLineSeparator()
    {
        var lexer = new Lexer();
        var context = MakeLexerContext("ls -al\r\nbb");

        var result = Process(lexer, context);

        var expectedTokens = new Token[]
        {
            new TextToken("ls -al"),
            Tokens.NewLine,
            new TextToken("bb"),
        };

        Assert.Equal(expectedTokens, result, TokenEqualityComparer.Instance);
    }
}

File: src/TextAnalysis.Tests/CommandLexerTests.cpp
#include "pch.h"
#include <vector>

#include <TextAnalysis/Buffer.h>
#include <TextAnalysis/BufferPosition.h>
#include <TextAnalysis/Lexer.h>

#include "TestMisc.h"

using namespace std;
using namespace Citron;

void RepeatLexCommand(vector<Token>* tokens, Lexer* lexer, int repeatCount)
{
    for (int i = 0; i < repeatCount; i++)
    {
        auto optLexResult = lexer->LexCommandMode();
        ASSERT_TRUE(optLexResult);
        
        tokens->push_back(std::move(optLexResult->token)); // ps
        *lexer = std::move(optLexResult->lexer);
    }
}

void RepeatLexNormal(vector<Token>* tokens, Lexer* lexer, bool bSkipNewLine, int repeatCount)
{
    for (int i = 0; i < repeatCount; i++)
    {
        auto optLexResult = lexer->LexNormalMode(bSkipNewLine);
        ASSERT_TRUE(optLexResult);

        tokens->push_back(std::move(optLexResult->token)); // ps
        *lexer = std::move(optLexResult->lexer);
    }
}

vector<Token> Process(Lexer lexer)
{
    vector<Token> tokens;
    while (!lexer.IsReachedEnd())
    {
        auto optLexResult = lexer.LexCommandMode();
        if (!optLexResult) break;

        tokens.push_back(std::move(optLexResult->token));
        lexer = std::move(optLexResult->lexer);
    }

    return tokens;
}

TEST(CommandLexer, ProcessStringExpInCommandMode)
{   
    auto [buffer, lexer] = Prepare(U"  p$$s${ ccc } \"ddd $e  \r\n }"s);

    vector<Token> tokens;

    RepeatLexCommand(&tokens, &lexer, 2);
    RepeatLexNormal(&tokens, &lexer, false, 2);
    RepeatLexCommand(&tokens, &lexer, 6);

    vector<Token> expectedTokens = {
        TextToken("  p$s"),
        DollarLBraceToken(),
        IdentifierToken("ccc"),
        RBraceToken(),
        TextToken(" \"ddd "),
        IdentifierToken("e"),
        TextToken("  "),
        NewLineToken(),
        TextToken(" "),
        RBraceToken(),
    };

    EXPECT_EQ(tokens, expectedTokens);
}

TEST(CommandLexer, LexCommands)
{
    auto [buffer, lexer] = Prepare(U"ls -al");
    auto tokens = Process(lexer);

    vector<Token> expectedTokens = {
        TextToken("ls -al")
    };

    EXPECT_EQ(tokens, expectedTokens);
}

TEST(CommandLexer, LexMultiLines)
{
    auto [buffer, lexer] = Prepare(UR"---(
hello world \n

    hello    

})---"s);

    vector<Token> tokens;
    RepeatLexCommand(&tokens, &lexer, 6);

    vector<Token> expected = {
        NewLineToken(),
        TextToken("hello world \\n"), NewLineToken(), // skip multi newlines
        TextToken("    hello    "), NewLineToken(),
        RBraceToken()
    };

    EXPECT_EQ(tokens, expected);
}

TEST(CommandLexer, LexCommandsWithLineSeparator)
{
    auto [buffer, lexer] = Prepare(U"ls -al\r\nbb"s);
    auto tokens = Process(lexer);

    vector<Token> expected = {
        TextToken("ls -al"),
        NewLineToken(),
        TextToken("bb"),
    };

    EXPECT_EQ(tokens, expected);
}